### PR TITLE
Merge TFS 8193011: Constrained language mode doesn't allow CIM cmdlets

### PR DESCRIPTION
**Pull request by PaulHigin**
Some CIM needed types were not included in the PowerShell core types white list preventing the CimCmdlets module commands from running in a ConstrainedLanguage PowerShell.

Fix is to add the missing types to the white list.  Fix also includes modifying the dot source check to allow script that only implements PowerShell commands, which allows the CimCmdlets module to load in ConstrainedLanguage mode.

### CODE REVIEW

File: src/System.Management.Automation/engine/CommandProcessorBase.cs [hunk 1]
```diff
@@ -169,24 +169,38 @@ internal bool UseLocalScope
             InvocationInfo invocationInfo)
         {
             // If we are in a constrained language mode (Core or Restricted), block it.
-            // This goes both ways:
+            // We are currently restricting in one direction:
             //    - Can't dot something from a more permissive mode, since that would probably expose
             //      functions that were never designed to handle untrusted data.
-            //    - Can't dot something from a less permissive mode, since that might introduce tainted
-            //      data into the current scope.
```
Comment (daxian-dbw) [inline src/System.Management.Automation/engine/CommandProcessorBase.cs]: This comment is a little confusing. By looking at the code, the script block may from a more permissive mode (FullLanguage) or a less permissive mode (NoLanguage), but as long as it's just powershell commands plus parameters then it's allowed to run. So it seems there is really no one direction restriction.

File: src/System.Management.Automation/engine/parser/TypeResolver.cs [hunk 2]
```diff
@@ -716,9 +722,13 @@ internal static class CoreTypes
                     { typeof(void),                                        new[] { "void" } },
                     { typeof(IPAddress),                                   new[] { "ipaddress" } },
                     { typeof(DscLocalConfigurationManagerAttribute),       new[] {"DscLocalConfigurationManager"}},
+                    { typeof(WildcardPattern),                             new[] { "WildcardPattern" } },
+                    { typeof(X509Certificate),                             new[] { "X509Certificate" } },
+                    { typeof(X500DistinguishedName),                       new[] { "X500DistinguishedName" } },
                     { typeof(XmlDocument),                                 new[] { "xml" } },
+                    { typeof(CimSession),                                  new[] { "CimSession" } },
```
Comment (daxian-dbw) [inline src/System.Management.Automation/engine/parser/TypeResolver.cs]: >// Do not add types to this pool unless they can be safely exposed to an attacker and will not
// expose the ability to corrupt or escape PowerShell's environment. The following operations must
// be safe: type conversion, all constructors, all methods (instance and static), and
// and properties (instance and static).

Have you done a security review on those newly added types?

File: src/System.Management.Automation/engine/CommandProcessorBase.cs (hunk 1 above)
Comment (PaulHigin) [inline src/System.Management.Automation/engine/CommandProcessorBase.cs]: This code won't run for NoLanguage mode so the only direction checked is trusted (full language mode) script running in a constrained/restricted session.

File: src/System.Management.Automation/engine/parser/TypeResolver.cs (hunk 2 above)
Comment (PaulHigin) [inline src/System.Management.Automation/engine/parser/TypeResolver.cs]: Yes.  These changes have been reviewed and have gone through a security review with our security architect.

File: src/System.Management.Automation/engine/CommandProcessorBase.cs (hunk 1 above)
Comment (daxian-dbw) [inline src/System.Management.Automation/engine/CommandProcessorBase.cs]: >This code won't run for NoLanguage mode so the only direction checked is trusted (full language mode) script running in a constrained/restricted session.

I see. Can you please add this in the comment? I think that would be easier for people to understand the comment and the code.

**Comment by daxian-dbw** [comment]
LGTM.

**Pull request merged**